docs/testing-strategy.md

```diff
@@ -0,0 +1,152 @@
+isolation. They are intended to provide exhaustive tests of the code paths
+through a particular class or module through its public interface. As per
+the [test pyramid](https://martinfowler.com/bliki/TestPyramid.html) approach
+to software testing, these should provide the majority of test coverage.
+
+Characteristics of unit tests:
+
+- they should be concerned with the logic within the class/module being tested,
+  and not test logic defined elsewhere;
+- they may mock dependent objects and/or assert that particular external
+  methods are called;
+- they should test all code paths through a class/module;
+- they should test that side effects intended by the code occur, for example writing
+  to the database or making API calls.
+
+## Feature tests
+
+Feature tests are used to assert that Content Publisher behaves correctly from
+the perspective of someone using the application. This ensures we're testing
+the system in a way representative of expected user behaviour. This makes the
+tests some of the most valuable of the application, however they are slower
+than other tests to run and can be difficult to debug. Therefore they are not
+intended to exhaustively test all the scenarios that can occur.
+
+Characteristics of feature tests:
+
+- they should test across multiple requests - consider a
+  [request test](#request-tests) to test a single request;
+- they should navigate the application using the same means as a user, for
+  example clicking links or submitting forms;
+- they shouldn't test dead ends in a user flow (such as validation or
+  permission issues); these are simpler tested in
+  [request tests](#request-tests);
+- incidental behaviour (for example flash messages) shouldn't be tested, this
+  can be tested at a higher granularity in [unit](#unit-tests) or
+  [request](#request-tests) tests;
+- they shouldn't test the effects a test has on the database, only user visible
+  signs of success should be asserted on;
+- they shouldn't mock application code;
+- there may be mocking and assertions that particular API calls are made to
+  external services;
+- they should be written in the [readable feature test
+  style][readable-feature-tests] popularised by Futurelearn.
+
+## Request tests
+
+The purpose of these tests is to test the HTTP API for the application from
+a machine perspective. They are used to determine that for a given input and
+application state a particular response is returned. These differ from feature
+tests in that they test a single endpoint at a time and can exhaust the logical
+outcomes of that endpoint where some situations may be difficult for
+a user to fall into.
+
+Characteristics of request tests:
+
+- they should test the variety of responses an endpoint returns, checking
+  aspects of the response such as status code and any flash messages;
+- they shouldn't test responses where there isn't specific logic written for
+  the scenario;
```

Comment:
This line doesn't really mean much to me, does this basically mean that we shouldn't use request tests if the controller action is simple and doesn't house any 'complex' logic?

Reply:
I meant it more about where Rails may do things automatically. In that we probably don't want request tests that test whether we get a 404 for every document that doesn't exist or a 400 when required parameters are missing.

I wanted it to be a little more nuanced that just don't test Rails exceptions as sometimes we'd want to test them when our code interacts with them - for example we might want to test we get a 404 when we try access an image for a document where the image exists but belongs to a different document.

Reply:
Oh okay I get you, maybe add an example like the ones you've stated to the bullet point? As without that context I feel it's quite hate to decipher what it actually means, at least to my brain! 😆

Reply:
Sure - I've added that in, hopefully it's worded ok wasn't as easy to phrase as I thought.

Reply:
Had a read through and it's much clearer! 🌟